cmd/convertor/builder/overlaybd_builder.go

```diff
@@ -127,6 +138,77 @@ func (e *overlaybdBuilderEngine) UploadImage(ctx context.Context) error {
 	return e.uploadManifestAndConfig(ctx)
 }
 
+func (e *overlaybdBuilderEngine) CheckForConvertedLayer(ctx context.Context, idx int, chainID string) (*specs.Descriptor, error) {
+	if e.db == nil {
+		return nil, errdefs.ErrNotFound
+	}
+
+	// try to find in the same repo then check existence on registry
+	entry := e.db.GetEntryForRepo(ctx, e.host, e.repository, chainID)
+	if entry != nil && entry.ChainID != "" {
+		desc := specs.Descriptor{
+			MediaType: e.mediaTypeImageLayer(),
+			Digest:    entry.ConvertedDigest,
+			Size:      entry.DataSize,
+		}
+		rc, err := e.fetcher.Fetch(ctx, desc)
+
+		if err == nil {
+			rc.Close()
+			logrus.Infof("layer %d found in remote with chainID %s", idx, chainID)
+			return &desc, nil
+		}
+		if errdefs.IsNotFound(err) {
+			// invalid record in db, which is not found in registry, remove it
+			err := e.db.DeleteEntry(ctx, e.host, e.repository, chainID)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// found record in other repos, try mounting it to the target repo
+	entries := e.db.GetCrossRepoEntries(ctx, e.host, chainID)
+	for _, entry := range entries {
+		desc := specs.Descriptor{
+			MediaType: e.mediaTypeImageLayer(),
+			Digest:    entry.ConvertedDigest,
+			Size:      entry.DataSize,
+			Annotations: map[string]string{
+				fmt.Sprintf("%s.%s", labelDistributionSource, e.host): entry.Repository,
+			},
+		}
+
+		_, err := e.pusher.Push(ctx, desc)
+		if errdefs.IsAlreadyExists(err) {
+			desc.Annotations = nil
+
+			if err := e.db.CreateEntry(ctx, e.host, e.repository, entry.ConvertedDigest, chainID, entry.DataSize); err != nil {
+				continue // try a different repo if available
+			}
+
+			logrus.Infof("layer %d mount from %s was successful", idx, entry.Repository)
+			logrus.Infof("layer %d found in remote with chainID %s", idx, chainID)
+			return &desc, nil
+		}
+	}
+
+	logrus.Infof("layer %d not found in remote", idx)
+	return nil, errdefs.ErrNotFound
+}
+
+func (e *overlaybdBuilderEngine) StoreConvertedLayerDetails(ctx context.Context, chainID string, idx int) error {
+	if e.db == nil {
+		return nil
+	}
+	return e.db.CreateEntry(ctx, e.host, e.repository, e.overlaybdLayers[idx].Digest, chainID, e.overlaybdLayers[idx].Size)
+}
+
+func (e *overlaybdBuilderEngine) DownloadConvertedLayer(ctx context.Context, idx int, desc *specs.Descriptor) error {
```

Comment:
double if using reference of specs makes sense

Reply:
Updated the functions of the interface to return the actual structs this is mostly to maintain consistency rather than because it changes anything.